File: data/two-pointers/two-sum.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> twoSum(vector<int>& numbers, int target) {
    int left = 0, right = numbers.size() - 1;  // Initialize two pointers

    while (left < right) {
        int total = numbers[left] + numbers[right];

        if (total == target)
            return {left + 1, right + 1};  // Return 1-indexed positions
        else if (total < target)
            left++;  // Move left pointer forward
        else
            right--;  // Move right pointer backward
    }

    return {};  // Should never reach here if there's always a valid solution
}

int main() {
    vector<int> numbers = {2, 7, 11, 15};
    int target = 9;

    vector<int> result = twoSum(numbers, target);
    cout << "[" << result[0] << ", " << result[1] << "]" << endl;  // Output: [1, 2]

    return 0;
}

File: data/two-pointers/two-sum.py
def twoSum(numbers, target):
    # Initialize two pointers:
    # 'left' starts at the beginning of the array (index 0)
    # 'right' starts at the end of the array (index len(numbers) - 1)
    left, right = 0, len(numbers) - 1  

    # Iterate as long as the 'left' pointer is less than the 'right' pointer
    while left < right:
        # Calculate the sum of the values at the 'left' and 'right' pointers
        total = numbers[left] + numbers[right]

        # Check if the sum of the two values is equal to the target
        if total == target:
            # If we find the correct sum, return their 1-indexed positions (left + 1 and right + 1)
            return [left + 1, right + 1]  

        # If the sum is less than the target, we need a larger sum, so move the left pointer forward
        elif total < target:
            left += 1  # Increment left to increase the sum (since numbers are sorted, the next number will be larger)

        # If the sum is greater than the target, we need a smaller sum, so move the right pointer backward
        else:
            right -= 1  # Decrement right to decrease the sum (since numbers are sorted, the next number will be smaller)

    # If we reach here, it means no valid solution was found, but the problem guarantees a solution exists.
    # So, this return statement is never actually reached in a valid problem setup.
    return []  


# Example usage:
numbers = [2, 7, 11, 15]
target = 9
result = twoSum(numbers, target)
print(result)  # Output: [1, 2]
